#### Objective

##### Predict whether dresses reviews are positive (>3 stars) or neutral/negative (<4 stars). Write a Jupyter Notebook report documenting your investigation.

### Explain briefly in your own words how the bag-of-words model and Naïve Bayes work, and how they work together.


## The bag-of-words model:
#### First the data is taken and the words a tokenized
#### Then the stop words are filterd out, this are words that are just used alot in a language so they are not really distinctive
#### The remaining tokens get stempt or Lemmatization 
#### stemming means that you take all words and you take the end of the word 
#### Lemmatization makes you go back to the grammatical stem of the word

#### after this you can try to quantify the words
#### the bag of words then method then  counts all instances of all the tokens in a given observation
#### Example : 'The cat jumped out of the hat' so first the frequent words get removed.
#### you will and up keeping the words: cat, jumped, out , hat. all of these 4 words occure once 
#### so you will get.  cat 1, jumped 1, out 1, hat 1
#### so every the times these words a used in a dataset they are counted 

## Naive Bayes work
#### This method looks at the propebility that, for example, if a mail is spam or not. Each word is then divided into true or false. So with the bag of words methode the words can get labelt. For exmaple the word viagra has a really high chance to be spam. the propebility of the mail not being spam might be 0.001% so it probobly spam.


In [5]:
import pandas as pd

## Pre-processing steps (don’t forget to filter out all non-dress reviews). The head() resulting dataframe


In [6]:
df = pd.read_csv('Assignment7.csv')
df = df.loc[(df['Department Name'] == 'Dresses')]
df.head()

Unnamed: 0  Clothing ID  Age                    Title  \
1           1         1080   34                      NaN   
2           2         1077   60  Some major design flaws   
5           5         1080   49  Not for the very petite   
8           8         1077   24               Flattering   
9           9         1077   34        Such a fun dress!   

                                         Review Text  Rating  Recommended IND  \
1  Love this dress!  it's sooo pretty.  i happene...       5                1   
2  I had such high hopes for this dress and reall...       3                0   
5  I love tracy reese dresses, but this one is no...       2                0   
8  I love this dress. i usually get an xs but it ...       5                1   
9  I'm 5"5' and 125 lbs. i ordered the s petite t...       5                1   

   Positive Feedback Count Division Name Department Name Class Name  
1                        4       General         Dresses    Dresses  
2                        0       General         Dresses    Dresses  
5                        4       General         Dresses    Dresses  
8                        0       General         Dresses    Dresses  
9                        0       General         Dresses    Dresses

### Text pre-processing steps resulting in a document-feature matrix

In [7]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.naive_bayes import MultinomialNB

In [8]:
df['Department Name'].value_counts()

Dresses    6319
Name: Department Name, dtype: int64

In [9]:
from sklearn.feature_extraction.text import CountVectorizer #The CountVectorizer object

text = df['Review Text'].values.astype('U') #Taking the text from the df. We need to convert it to Unicode

vect = CountVectorizer(stop_words='english') #Create the CV object, with English stop words
vect = vect.fit(text) #We fit the model with the words from the review text
vect
feature_names = vect.get_feature_names() #Get the words from the vocabulary
#feature_names
print(f"There are {len(feature_names)} words in the vocabulary. A selection: {feature_names[500:520]}")

There are 8080 words in the vocabulary. A selection: ['airier', 'airiness', 'airism', 'airline', 'airplane', 'airplanes', 'airport', 'airy', 'aize', 'aka', 'akward', 'al', 'alas', 'albeit', 'alerations', 'alert', 'alexandria', 'align', 'aligned', 'alignment']


#### the dictionary consist of 8080 words, we can count the occurences of each word for each review. This way, we can create a document-feature matrix, with documents (reviews) in the rows, and features (words) in the columns.

In [10]:
docu_feat = vect.transform(text) #The transform method from the CountVectorizer object creates the matrix
print(docu_feat[0:100,0:100]) #Let's print a little part of the matrix: the first 50 words & documents

  (2, 8)	1
  (4, 72)	1
  (19, 80)	1
  (20, 38)	1
  (21, 4)	1
  (21, 45)	1
  (22, 12)	1
  (25, 55)	1
  (26, 40)	1
  (28, 60)	1
  (35, 12)	2
  (35, 59)	3
  (38, 86)	1
  (39, 31)	1
  (39, 96)	1
  (49, 91)	1
  (57, 72)	1
  (58, 0)	1
  (62, 59)	4
  (68, 12)	1
  (68, 59)	1
  (69, 72)	1
  (75, 45)	1
  (77, 33)	1
  (78, 83)	1
  (80, 12)	1
  (98, 70)	1
  (99, 12)	1
  (99, 59)	1


#### Train a Naïve Bayes classifier predicting whether a review is positive (>3 stars) or neutral/negative (<4 stars).


In [25]:
df['Goodrating'] = df['Rating'] > 3 
df.head(30)

Unnamed: 0  Clothing ID  Age  \
1             1         1080   34   
2             2         1077   60   
5             5         1080   49   
8             8         1077   24   
9             9         1077   34   
10           10         1077   53   
11           11         1095   39   
12           12         1095   53   
14           14         1077   50   
19           19         1077   47   
21           21         1080   55   
22           22         1077   31   
23           23         1077   34   
52           52         1104   39   
58           58         1095   36   
60           60         1095   41   
69           69         1078   56   
74           74         1081   59   
76           76         1082   31   
79           79         1081   62   
80           80         1072   68   
87           87         1081   39   
90           90         1078   51   
93           93         1081   31   
109         109         1081   47   
111         111         1081   44   
113         113         1081   61   
114         114         1082   32   
116         116         1072   51   
117         117         1078   32   

                                            Title  \
1                                             NaN   
2                         Some major design flaws   
5                         Not for the very petite   
8                                      Flattering   
9                               Such a fun dress!   
10   Dress looks like it's made of cheap material   
11                                            NaN   
12                                     Perfect!!!   
14            Pretty party dress with some issues   
19                        Stylish and comfortable   
21                                      I'm torn!   
22                         Not what it looks like   
23                    Like it, but don't love it.   
52                                            NaN   
58                                            NaN   
60                                Gorgeous dress!   
69                           Great summer fabric!   
74                                            NaN   
76                    Runs large in top and waist   
79                      So cute...but dry clean?!   
80                          Elegant classic dress   
87                          True to size to large   
90                         Sweet flattering dress   
93                                            NaN   
109                                           NaN   
111                           An almost for me...   
113                                Cute and comfy   
114                         Disappointing quality   
116                             Very vintage feel   
117                                           NaN   

                                           Review Text  Rating  \
1    Love this dress!  it's sooo pretty.  i happene...       5   
2    I had such high hopes for this dress and reall...       3   
5    I love tracy reese dresses, but this one is no...       2   
8    I love this dress. i usually get an xs but it ...       5   
9    I'm 5"5' and 125 lbs. i ordered the s petite t...       5   
10   Dress runs small esp where the zipper area run...       3   
11   This dress is perfection! so pretty and flatte...       5   
12   More and more i find myself reliant on the rev...       5   
14   This is a nice choice for holiday gatherings. ...       3   
19   I love the look and feel of this tulle dress. ...       5   
21   I'm upset because for the price of the dress, ...       4   
22   First of all, this is not pullover styling. th...       2   
23   Cute little dress fits tts. it is a little hig...       3   
52   Love the color and style, but material snags e...       3   
58   I got this in the petite length, size o, and i...       4   
60   This dress is simply beautiful and stunning. i...       5   
69   I really wanted this to work. alas, it had a s...       3   
74   I passed up this dre

1    False
2     True
5     True
8    False
9    False
Name: bad, dtype: bool

### Split the file into a training and a test set.


In [12]:
y = df['Goodrating'] #the positive rating (above 3)becomes the Y-variable
X = docu_feat 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1) #split the data, store it into different variables

In [13]:
nb = MultinomialNB() 
nb = nb.fit(X_train, y_train)

### Evaluate the performance of the model on the test set.¶


In [14]:
y_test_p = nb.predict(X_test)
nb.score(X_test, y_test)

0.8554852320675106

### The model predicts with 85.55%. accuracy. their are 5 possibilities because you can rate from 1-5.  the model calculates fairly precisely

In [17]:
print(df.iloc[0,4]) # takes the content from the 5th row 
print(nb.predict_proba(X[0])) # where the prediction = False 

Love this dress!  it's sooo pretty.  i happened to find it in a store, and i'm glad i did bc i never would have ordered it online bc it's petite.  i bought a petite and am 5'8".  i love the length on me- hits just a little below the knee.  would definitely be a true midi on someone who is truly petite.
[[5.33741270e-05 9.99946626e-01]]


In [27]:
for i in range(25):
    prob = nb.predict_proba(X[i])
    print(f"Review: {i}. {df.iloc[i,4]}")
    print(f"False: {prob[0,0]}, Positive: {prob[0,1]}")

Review: 0. Love this dress!  it's sooo pretty.  i happened to find it in a store, and i'm glad i did bc i never would have ordered it online bc it's petite.  i bought a petite and am 5'8".  i love the length on me- hits just a little below the knee.  would definitely be a true midi on someone who is truly petite.
False: 5.337412695289174e-05, Positive: 0.9999466258730467
Review: 1. I had such high hopes for this dress and really wanted it to work for me. i initially ordered the petite small (my usual size) but i found this to be outrageously small. so small in fact that i could not zip it up! i reordered it in petite medium, which was just ok. overall, the top half was comfortable and fit nicely, but the bottom half had a very tight under layer and several somewhat cheap (net) over layers. imo, a major design flaw was the net over layer sewn directly into the zipper - it c
False: 0.9999990978503296, Positive: 9.021496679830375e-07
Review: 2. I love tracy reese dresses, but this one is 

### I dont konw how to filter the wrong predicted reviews
